geospaas_harvesting/crawlers.py

```diff
@@ -443,17 +445,45 @@ def set_initial_state(self):
         """
         self._urls = []
         self._to_process = [self.root_url.path or '/']
+        self.connect()
+
+    def connect(self):
+        """Creates an FTP connection and logs in"""
         self.ftp = ftplib.FTP(self.root_url.netloc, user=self.username, passwd=self.password)
         try:
             self.ftp.login(self.username, self.password)
         except ftplib.error_perm as err_content:
-            # these two cases are in the mentioned FTP servers that deals with "login once again"
+            # these errors happen when we try to log in twice, so they can be ignored
             if not (err_content.args[0].startswith('503') or err_content.args[0].startswith('230')):
                 raise
 
+    class Decorators():
+        """Decorators for the FTPCrawler"""
+        @staticmethod
+        def retry_on_timeout(method):
+            """Decorator which re-creates the FTP connection if a timeout error occurs"""
+            @functools.wraps(method)
+            def wrapper_reconnect(crawler_instance, *args, **kwargs):
+                """Try to execute the decorated method.
+                If a 421 error occurs, re-create the connection
+                and try to run the method again (once).
+                """
+                try:
+                    return method(crawler_instance, *args, **kwargs)
+                except (ftplib.error_temp, ConnectionError) as error:
+                    if isinstance(error, ftplib.error_temp) and '421' not in error.args[0]:
+                        raise
```

Comment:
Shall it not raise something more specific? E.g. `error` or at least `ConnectionError` so that it can be caught better upstream?

Reply:
Using `raise` without argument re-raises the caught exception. It's equivalent to `raise error`.